### PR TITLE
capture: improve typing

### PR DESCRIPTION
**Pull request by bluetech**
I'm looking at improving the capture plugin again, but first I thought I'd finish the typing. Adds some verbosity but needed to "capture" the current usage.

Might be easier to review commit by commit.

### CODE REVIEW

File: src/_pytest/capture.py [hunk 1]
```diff
@@ -185,19 +190,27 @@ def write(self, s: str) -> int:
...
 
-class DontReadFromInput:
-    encoding = None
+class DontReadFromInput(TextIO):
+    @property
+    def encoding(self) -> str:
+        return sys.__stdin__.encoding
 
-    def read(self, *args):
+    def read(self, size: int = -1, /) -> str:
```
Comment (nicoddemus) [inline src/_pytest/capture.py:200]: positional-only parameters are not supported in Python 3.7.

File: src/_pytest/capture.py [hunk 2]
```diff
@@ -185,19 +190,27 @@ def write(self, s: str) -> int:
...
 
     readline = read
-    readlines = read
-    __next__ = read
 
-    def __iter__(self):
+    def __next__(self) -> str:
+        return self.readline()
+
+    def readlines(self, hint: Optional[int] = -1, /) -> list[str]:
```
Comment (nicoddemus) [inline src/_pytest/capture.py:210]: `List[str]` needed, or `from __future__ import annotations` but not sure it works in 3.7.

**Review by nicoddemus — APPROVED** [review]
LGTM!

Just have to adapt some of the unsupported features for Python 3.8/3.7. 👍

**Comment by bluetech** [comment]
@nicoddemus Thanks for the notes, fixed. Also fixed the coverage.

**Pull request merged**